fabtests/common/shared.c

```diff
@@ -2060,25 +2060,60 @@ ssize_t ft_inject(struct fid_ep *ep, fi_addr_t fi_addr, size_t size)
 	return ret;
 }
 
-ssize_t ft_post_rma(enum ft_rma_opcodes op, struct fid_ep *ep, size_t size,
+/**
+ * @brief calculate remote's addr for tx_buf/rx_buf operations
+ *
+ * buf must be a pointer within tx_buf or rx_buf.
+ *
+ * buf is found either to be in the tx_buf or rx_buf, and addr is
+ * computed to be in the opposite buffer, at the same offset from
+ * the start.
+ *
+ * @return addr_offset, relative to remote's rx_buf MR.
+ */
+static uint64_t ft_rma_addr_from_buf(char *buf) {
+	ssize_t offset;
+	size_t buf_bytes = tx_buf - rx_buf;
+
+	/**
+	 * note buffers are adjacent: [rx_buf-------|tx_buf--------]
+	 * and buffers are equally sized. */
+	assert(buf >= rx_buf && buf < tx_buf + buf_bytes);
+
+	if (buf >= tx_buf) {
+		/* local buf is within tx_buf */
+		offset = buf - tx_buf;
+		/* remote buf is within rx_buf */
+		return offset;
+	} else {
+		/* local buf is within rx_buf */
+		offset = buf - rx_buf;
+		/* remote buf is within tx_buf */
+		return offset + buf_bytes;
+	}
+}
+
+ssize_t ft_post_rma(enum ft_rma_opcodes op, char *buf, size_t size,
 		struct fi_rma_iov *remote, void *context)
 {
+	uint64_t addr = remote->addr + ft_rma_addr_from_buf( buf );
+
 	switch (op) {
 	case FT_RMA_WRITE:
 		FT_POST(fi_write, ft_progress, txcq, tx_seq, &tx_cq_cntr,
-			"fi_write", ep, tx_buf, size, mr_desc,
-			remote_fi_addr, remote->addr, remote->key, context);
+			"fi_write", ep, buf, size, mr_desc,
+			remote_fi_addr, addr, remote->key, context);
```

Comment:
I would replace the proposed ft_rma_addr_from_buf() with separate calls.  Example:
```
static size_t ft_write_offset(char *buf)
{
    assert(buf >= tx_buf && buf < (tx_buf + buf_size / 2));
    return buf - tx_buf;
}

/* Reads copy data from the remote tx_buf area into the local rx_buf area */
static size_t ft_read_offset(char *buf)
{
    assert(buf >= rx_buf && buf < (rx_buf + buf_size / 2));
    /* tx_buf area is the latter half of the remote region */
    return buf - rx_buf + buf_size / 2;
}
```

Reply:
Done.